cmd/agent/app/flags.go

```diff
@@ -22,69 +22,83 @@ package app
 
 import (
 	"flag"
-	"fmt"
 	"strings"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	suffixWorkers             = "workers"
+	suffixServerQueueSize     = "server-queue-size"
+	suffixServerMaxPacketSize = "server-max-packet-size"
+	suffixServerHostPort      = "server-host-port"
+	collectorHostPort         = "collector.host-port"
+	httpServerHostPort        = "http-server.host-port"
+	discoveryMinPeers         = "discovery.min-peers"
 )
 
-// Bind binds the agent builder to command line options
-func (b *Builder) Bind(flags *flag.FlagSet) {
+// CreateFlags for this builder based on the application's parameters
+func (b *Builder) CreateFlags() *flag.FlagSet {
+	flagSet := &flag.FlagSet{}
+
 	for i := range b.Processors {
 		p := &b.Processors[i]
-		name := "processor." + string(p.Model) + "-" + string(p.Protocol) + "."
-		flags.IntVar(
-			&p.Workers,
-			name+"workers",
+		prefix := flagPrefix(p)
+		flagSet.Int(
+			prefix+suffixWorkers,
 			p.Workers,
 			"how many workers the processor should run")
-		flags.IntVar(
-			&p.Server.QueueSize,
-			name+"server-queue-size",
+		flagSet.Int(
+			prefix+suffixServerQueueSize,
 			p.Server.QueueSize,
 			"length of the queue for the UDP server")
-		flags.IntVar(
-			&p.Server.MaxPacketSize,
-			name+"server-max-packet-size",
+		flagSet.Int(
+			prefix+suffixServerMaxPacketSize,
 			p.Server.MaxPacketSize,
 			"max packet size for the UDP server")
-		flags.StringVar(
-			&p.Server.HostPort,
-			name+"server-host-port",
+		flagSet.String(
+			prefix+suffixServerHostPort,
 			p.Server.HostPort,
 			"host:port for the UDP server")
 	}
-	flags.Var(
-		&stringSliceFlag{slice: &b.CollectorHostPorts},
-		"collector.host-port",
+	flagSet.String(
+		collectorHostPort,
+		strings.Join(b.CollectorHostPorts, ","),
 		"comma-separated string representing host:ports of a static list of collectors to connect to directly (e.g. when not using service discovery)")
-	flags.StringVar(
-		&b.HTTPServer.HostPort,
-		"http-server.host-port",
+	flagSet.String(
+		httpServerHostPort,
 		b.HTTPServer.HostPort,
-		"host:port of the http server (e.g. for /sampling and /baggage endpoint)")
-	flags.IntVar(
-		&b.DiscoveryMinPeers,
-		"discovery.min-peers",
+		"host:port of the http server (e.g. for /sampling point and /baggage endpoint)")
+	flagSet.Int(
+		discoveryMinPeers,
 		3,
 		"if using service discovery, the min number of connections to maintain to the backend")
-}
 
-type stringSliceFlag struct {
-	slice *[]string
+	return flagSet
 }
 
-// String formats the flag's value, part of the flag.Value interface.
-func (c *stringSliceFlag) String() string {
-	return fmt.Sprint(c.slice)
-}
+// Initialize builder properties
+func (b *Builder) Initialize() {
+	// make sure that viper is initialized with default values for a case when CreateFlags is not
+	// passed to viper
+	b.CreateFlags().VisitAll(func(flag *flag.Flag) {
+		viper.SetDefault(flag.Name, flag.DefValue)
```

Comment:
this still concerns me. Can we do away with global vars?  All Viper's functionality is accessible via an instance, it does not need to be global.